oauthenticator/globus.py

```diff
@@ -0,0 +1,101 @@
+"""
+Custom Authenticator to use Globus OAuth2 with JupyterHub
+"""
+import pwd
+from tornado import gen, web
+from tornado.auth import OAuth2Mixin
+from tornado.concurrent import return_future
+from tornado.web import HTTPError
+from .oauth2 import OAuthLoginHandler, OAuthenticator
+
+try:
+    import globus_sdk
+except:
+    raise ImportError("Trying to use the Globus Auth "
+                      "authenticator, but globus_sdk "
+                      "is not installed")
+
+
+class GlobusMixin(OAuth2Mixin):
+    """
+    Overriding the tornado function because
+    globus provides it's own method to assemble the
+    authorize url
+    """
+    @return_future
+    def authorize_redirect(self, client=None, callback=None):
+        self.redirect(client.oauth2_get_authorize_url())
+        callback()
+
+
+class GlobusLoginHandler(OAuthLoginHandler, GlobusMixin):
+    def get(self):
+        # Doing the scope weirdness for the globus_sdk
+        scopes = ['openid', 'profile', 'email',
+                  'urn:globus:auth:scope:auth.globus.org:view_identities']
+        scope_string = (' ').join(scopes)
+        redirect_uri = self.authenticator.get_callback_url(self)
+        client = self.authenticator.globus_portal_client()
+        client.oauth2_start_flow(
+            redirect_uri,
+            requested_scopes=scope_string,
+            refresh_tokens=True)
+        self.log.info('globus redirect: %r', redirect_uri)
+        self.authorize_redirect(client)
+
+
+class GlobusOAuthenticator(OAuthenticator):
+
+    login_service = "Globus"
+    login_handler = GlobusLoginHandler
+
+    def globus_portal_client(self):
```

Comment:
SHould this be a property instead of a method? Is there a reason to create a new Client on every request?